hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/TypeEquivalance.java

```diff
@@ -41,6 +52,12 @@ public boolean isComparable(Object o) {
 
 	@Override
 	public int compare(Object x, Object y) {
-		return type.compare(x, y);
+		Class<?> typeClass = type.getReturnedClass();
+		if (typeClass.isInstance(x) && typeClass.isInstance(y)) {
+			return type.compare(x, y);
+		}
+		else {
+			return 0;
```

Comment:
`CollectionType.compare` has
```
	@Override
	public int compare(Object x, Object y) {
		return 0; // collections cannot be compared
	}
``` 
and `EntityType` has
```
        @Override
	public int compare(Object x, Object y) {
		return 0; //TODO: entities CAN be compared, by PK, fix this! -> only if/when we can extract the id values....
	}
```
so I've used 'error handling' consistent with these.